File: sdk/python/meerkat_mobkit/config/event_log.py
"""Event-log durability declaration for MobKit runtime.

The gateway ingests operational events only when
``runtime_options.event_log`` is configured; the absence of the key means
no ingestion at all (the honest silent case). The two wire-supported
storage kinds are both *declared ephemeral* choices:

- :func:`memory` — a bounded, queryable in-process store
  (``{"storage": "memory"}``); serves ``mobkit/query_events``.
- :func:`null` — events are explicitly dropped (``{"storage": "null"}``);
  queries return empty.

Durable event-log backends are embedder-only today: the Rust
``UnifiedRuntimeBuilder`` accepts any ``EventLogStore`` implementation, but
the gateway wire supports only the declarations above and rejects anything
else at startup (``unsupported runtime_options.event_log.storage``). The
resolved choice is visible in the storage census (``mobkit/status`` →
``storage.slots``).
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class MemoryEventLogConfig:
    """Bounded queryable in-process event store (declared ephemeral)."""

    batch_size: int | None = None
    flush_interval_ms: int | None = None

    def to_dict(self) -> dict[str, Any]:
        result: dict[str, Any] = {"storage": "memory"}
        if self.batch_size is not None:
            result["batch_size"] = self.batch_size
        if self.flush_interval_ms is not None:
            result["flush_interval_ms"] = self.flush_interval_ms
        return result


@dataclass(frozen=True)
class NullEventLogConfig:
    """Events are explicitly dropped (declared ephemeral, queries empty)."""

    def to_dict(self) -> dict[str, Any]:
        return {"storage": "null"}


def memory(
    batch_size: int | None = None,
    flush_interval_ms: int | None = None,
) -> MemoryEventLogConfig:
    """Declare a bounded queryable in-process event store."""
    return MemoryEventLogConfig(
        batch_size=batch_size,
        flush_interval_ms=flush_interval_ms,
    )


def null() -> NullEventLogConfig:
    """Declare that operational events are dropped."""
    return NullEventLogConfig()

File: sdk/typescript/src/config/event-log.ts
/**
 * Event-log durability declaration for MobKit runtime.
 *
 * The gateway ingests operational events only when
 * `runtime_options.event_log` is configured; the absence of the key means
 * no ingestion at all (the honest silent case). The two wire-supported
 * storage kinds are both *declared ephemeral* choices:
 *
 * - {@link memory} — a bounded, queryable in-process store
 *   (`{"storage": "memory"}`); serves `mobkit/query_events`.
 * - {@link nullStore} — events are explicitly dropped
 *   (`{"storage": "null"}`); queries return empty.
 *
 * Durable event-log backends are embedder-only today: the Rust
 * `UnifiedRuntimeBuilder` accepts any `EventLogStore` implementation, but
 * the gateway wire supports only the declarations above and rejects
 * anything else at startup. The resolved choice is visible in the storage
 * census (`mobkit/status` → `storage.slots`).
 */

/** A typed event-log declaration for `MobKitBuilder.eventLog()`. */
export interface EventLogDeclaration {
  toDict(): Record<string, unknown>;
}

/** Declare a bounded queryable in-process event store. */
export function memory(options?: {
  batchSize?: number;
  flushIntervalMs?: number;
}): EventLogDeclaration {
  // A zero interval panics the gateway's ingestion task
  // (`tokio::time::interval` requires a non-zero period), and the wire
  // only accepts positive integers.
  const flushIntervalMs = options?.flushIntervalMs;
  if (
    flushIntervalMs !== undefined &&
    (!Number.isInteger(flushIntervalMs) || flushIntervalMs <= 0)
  ) {
    throw new Error(
      `eventLog.memory: flushIntervalMs must be a positive integer (got ${flushIntervalMs})`,
    );
  }
  return {
    toDict() {
      const result: Record<string, unknown> = { storage: "memory" };
      if (options?.batchSize !== undefined) result.batch_size = options.batchSize;
      if (options?.flushIntervalMs !== undefined) {
        result.flush_interval_ms = options.flushIntervalMs;
      }
      return result;
    },
  };
}

/**
 * Declare that operational events are dropped. (Named `nullStore` because
 * `null` is a reserved word; re-exported as `eventLog.nullStore`.)
 */
export function nullStore(): EventLogDeclaration {
  return {
    toDict() {
      return { storage: "null" };
    },
  };
}
